dbms/src/Common/Spiller.cpp

```diff
@@ -0,0 +1,181 @@
+        ret = current_file_stream->block_in->read();
+        if (ret)
+            return ret;
+    }
+    return ret;
+}
+
+Block SpilledFilesInputStream::getHeader() const
+{
+    return header;
+}
+String SpilledFilesInputStream::getName() const
+{
+    return "SpilledFiles";
+}
+
+Spiller::Spiller(const String & id_, bool is_input_sorted_, size_t partition_num_, const String & spill_dir_, const Block & input_schema_, LoggerPtr logger_)
+    : id(id_)
+    , is_input_sorted(is_input_sorted_)
+    , partition_num(partition_num_)
+    , spill_dir(spill_dir_)
+    , input_schema(input_schema_)
+    , logger(logger_)
+{
+    if (spill_dir.at(spill_dir.size() - 1) != Poco::Path::separator())
+    {
+        spill_dir += Poco::Path::separator();
+    }
+    spilled_files.resize(partition_num);
+}
+
+bool Spiller::spillBlocks(const Blocks & blocks, size_t partition_id)
+{
+    RUNTIME_CHECK_MSG(partition_id < partition_num, "{}: partition id {} exceeds partition num {}.", id, partition_id, partition_num);
+    RUNTIME_CHECK_MSG(spill_finished == false, "{}: spill after the spiller is finished.", id);
+    /// todo append to existing file
+    auto spilled_file_name = nextSpillFileName(partition_id);
+    try
+    {
+        auto spilled_file = std::make_unique<SpilledFile>(spilled_file_name);
+        if (spilled_file->exists())
+            throw Exception("Duplicated spilled files, should not happens");
+        WriteBufferFromFile file_buf(spilled_file_name);
+        CompressedWriteBuffer compressed_buf(file_buf);
+        NativeBlockOutputStream block_out(compressed_buf, 0, blocks[0].cloneEmpty());
+        block_out.writePrefix();
+        for (const auto & block : blocks)
+        {
+            auto block_bytes_size = block.bytes();
+            block_out.write(block);
+            spilled_file->addSpilledDataSize(block_bytes_size);
+        }
+        {
+            std::lock_guard lock(spilled_files_mutex);
+            spilled_files[partition_id].emplace_back(std::move(spilled_file));
+        }
+        return true;
+    }
+    catch (...)
+    {
+        LOG_ERROR(logger, "Failed to spill block to disk for file {}, error message: {}", spilled_file_name, getCurrentExceptionMessage(false, false));
+        return false;
+    }
+}
+
+BlockInputStreams Spiller::restoreBlocks(size_t partition_id, size_t max_stream_size)
+{
+    RUNTIME_CHECK_MSG(partition_id < partition_num, "{}: partition id {} exceeds partition num {}.", id, partition_id, partition_num);
+    RUNTIME_CHECK_MSG(spill_finished, "{}: restore before the spiller is finished.", id);
+    if (is_input_sorted && spilled_files[partition_id].size() > max_stream_size)
+        LOG_WARNING(logger, "sorted spilled data restore does not take max_stream_size into account");
+    BlockInputStreams ret;
+    if (is_input_sorted)
+    {
+        for (const auto & file : spilled_files[partition_id])
+        {
+            if likely (file->exists())
```

Comment:
Whether to throw an exception or log if the file does not exist ？

Reply:
OK, I will add some logs.